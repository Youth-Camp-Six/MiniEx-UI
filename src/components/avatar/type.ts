import { ReactNode } from 'react';
type status = 'online' | 'processing' | 'away' | 'offline' | 'busy' | '';

export interface AvatarProps {
  className?: string;
  round?: string;
  status?: status;
  src?: string;
  alt?: string;
  width?: string;
  height?: string;
  statusPosition?: string;
  icon?: ReactNode;
  style?: React.CSSProperties;
  children?: ReactNode;
}
type baseOptions = {
  label: string;
  value: string;
};
export type Options = AvatarProps & baseOptions;
export interface AvatarGroupProps {
  className?: string;
  maxCount?: number;
  maxPopoverPlacement?: 'top' | 'bottom';
  maxPopoverTrigger?: 'hover' | 'focus' | 'click';
  maxStyle?: React.CSSProperties;
  options: Options[];
  width?: string;
  height?: string;
  style?: React.CSSProperties;
  children?: ReactNode;
}
