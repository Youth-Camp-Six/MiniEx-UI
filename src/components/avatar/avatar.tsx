import React, { memo, useEffect, useRef } from 'react';
import classNames from 'classnames';
import { AvatarProps } from './type';

export const Avatar: React.FC<AvatarProps> = (props) => {
  const {
    className,
    statusPosition,
    height,
    children,
    src,
    style,
    alt,
    status,
    icon,
    width,
    round,
    ...restProps
  } = props;

  const classesAva = classNames('mi-avatar', className);
  const classesImg = classNames({
    [`mi-avatar-${round}`]: round,
  });
  const classDot = classNames('dot', {
    [`dot-${status}`]: status,
    [`dot-${statusPosition}`]: statusPosition,
  });

  const avatarNodeRef = useRef<HTMLSpanElement>(null);
  const [scale, setScale] = React.useState(1);
  useEffect(() => {
    if (width && avatarNodeRef.current && avatarNodeRef.current.offsetWidth > parseInt(width)) {
      setScale(parseInt(width) / avatarNodeRef.current.offsetWidth - 0.1);
    }
    //eslint-disable-next-line
  }, [])
  return (
    <div className={classesAva} style={{ width: `${width}`, height: `${height}` }} {...restProps}>
      {src ? (
        <img src={src} alt={alt} className={classesImg} />
      ) : (
        <div className='icon-box' style={style}>
          {icon ? (
            icon
          ) : (
            <span style={{ transform: `scale(${scale})`, fontSize: '18px' }} ref={avatarNodeRef}>
              {children}
            </span>
          )}
        </div>
      )}
      {status ? <span className={classDot}></span> : null}
    </div>
  );
};

Avatar.defaultProps = {
  round: '',
  src: '',
  status: '',
  statusPosition: '',
  height: '40px',
  width: '40px',
  alt: '',
};

export default memo(Avatar);
