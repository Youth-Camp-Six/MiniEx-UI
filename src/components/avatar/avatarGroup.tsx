import React, { memo, useEffect, useState } from 'react';
import { AvatarGroupProps, Options } from './type';
import { Flex, Avatar, Popover } from '../../index';

export const AvatarGroup: React.FC<AvatarGroupProps> = (props) => {
  const {
    className,
    maxCount,
    options,
    maxPopoverPlacement,
    maxPopoverTrigger,
    maxStyle,
    width,
    height,
    style,
    ...restProps
  } = props;

  const numOfChildren = options.length;
  const [childrenShow, setChildrenShow] = useState<Options[]>([]);
  const [childrenHidden, setChildrenHidden] = useState<Options[]>([]);

  const HiddenBody = (arr: Options[]) => {
    return (
      <Flex x='start' gap={0}>
        {arr.map((item) => {
          return (
            <Avatar
              className={item.className}
              icon={item.icon}
              round={item.round}
              key={item.value}
              status={item.status}
              src={item.src}
              alt={item.alt}
              statusPosition={item.statusPosition}
              width={width}
              height={height}
              style={item.style}
            >
              {item.label}
            </Avatar>
          );
        })}
      </Flex>
    );
  };

  useEffect(() => {
    if (maxCount && maxCount < numOfChildren) {
      setChildrenShow(options.slice(0, maxCount));
      setChildrenHidden(options.slice(maxCount, numOfChildren));
    } else {
      setChildrenShow(options);
    }
    //eslint-disable-next-line
  }, [options])
  return (
    <Flex x='start' gap={0} className={className} style={style} {...restProps}>
      {childrenShow.map((item, index) => {
        return (
          <Avatar
            className={`${index !== 0 ? 'group-item ' : ''}` + item.className}
            icon={item.icon}
            round={item.round}
            key={item.value}
            status={item.status}
            src={item.src}
            alt={item.alt}
            statusPosition={item.statusPosition}
            width={width}
            height={height}
            style={item.style}
          >
            {item.label}
          </Avatar>
        );
      })}
      {maxCount && maxCount < options.length ? (
        <Popover
          trigger={maxPopoverTrigger}
          placement={maxPopoverPlacement}
          childrenData={HiddenBody(childrenHidden)}
        >
          <Avatar className='group-item' style={maxStyle}>{`+${numOfChildren - maxCount}`}</Avatar>
        </Popover>
      ) : null}
    </Flex>
  );
};

AvatarGroup.defaultProps = {
  maxPopoverPlacement: 'bottom',
  maxPopoverTrigger: 'hover',
};

export default memo(AvatarGroup);
