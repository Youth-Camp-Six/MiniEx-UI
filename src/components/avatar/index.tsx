import Avatar from './avatar';

export * from './type';
export { Avatar };
