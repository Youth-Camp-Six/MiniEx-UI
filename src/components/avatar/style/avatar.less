.mi-avatar {
  position: relative;
  display: inline-block;

  img {
    width: 100%;
    height: 100%;
    background-color: var(--mi-bgcolor-2);
    border-radius: var(--mi-border-radius);
  }

  .mi-avatar-round {
    border-radius: 50%;
  }

  .dot {
    position: absolute;
    top: -0.3rem;
    left: -0.3rem;
    z-index: 99;
    width: 0.6rem;
    height: 0.6rem;
    border: 2px var(--mi-white-color) solid;
    border-radius: 50%;
    content: '';
  }

  .dot-online {
    background-color: var(--mi-success-color);
  }

  .dot-busy {
    background-color: var(--mi-error-color);
  }

  .dot-offline {
    background-color: var(--mi-normal-color-dark);
  }

  .dot-processing {
    background-color: var(--mi-info-color);
  }

  .dot-away {
    background-color: var(--mi-warning-color);
  }

  .dot-top-left {
    top: -0.25rem;
    left: -0.25rem;
  }

  .dot-top-right {
    top: -0.25rem;
    right: -0.25rem;
    left: auto;
  }

  .dot-bottom-left {
    top: auto;
    bottom: -0.25rem;
    left: -0.25rem;
  }

  .dot-bottom-right {
    top: auto;
    right: -0.25rem;
    bottom: -0.25rem;
    left: auto;
  }

  .icon-box {
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    color: #fff;
    font-size: 14px;
    line-height: 30px;
    white-space: nowrap;
    text-align: center;
    vertical-align: middle;
    list-style: none;
    background: rgb(0 0 0 / 25%);
    border: 1px solid transparent;
    border-radius: 50%;
  }
}

.group-item {
  margin-inline-start: -8px;
  border-color: #fff;
}
