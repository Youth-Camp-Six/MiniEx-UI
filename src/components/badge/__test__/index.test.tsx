/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { Badge } from '..';
import { Avatar } from '../../avatar';

describe('Badge', () => {
  // 快照测试
  it('test case1 for direct use', () => {
    const wrapper = render(
      <div>
        <Badge value='5' />
        <Badge value='15' type='info' />
        <Badge value='65' type='success' />
        <Badge value='9' type='warning' />
        <Badge value='99+' type='error' />
        <Badge round='round' />
        <Badge round='round' type='info' />
        <Badge round='round' type='success' />
        <Badge round='round' type='warning' />
        <Badge round='round' type='error' />
      </div>
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });
  it('test case1 for avatar', () => {
    const wrapper = render(
      <div>
        <Badge value='9' type='info'>
          <Avatar>U</Avatar>
        </Badge>
      </div>
    );
    expect(wrapper.container.firstChild).toMatchSnapshot();
  });
});
