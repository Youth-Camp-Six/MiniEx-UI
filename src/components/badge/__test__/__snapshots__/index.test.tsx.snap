// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge test case1 for avatar 1`] = `
<div>
  <div
    class="mi-badge mi-badge-info"
  >
    <div
      class="mi-badge-value"
    >
      9
    </div>
    <div
      class="mi-avatar"
      style="width: 40px; height: 40px;"
    >
      <div
        class="icon-box"
      >
        <span
          style="transform: scale(1); font-size: 18px;"
        >
          U
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`Badge test case1 for direct use 1`] = `
<div>
  <div
    class="mi-badge mi-badge-normal"
  >
    <div
      class="mi-badge-value"
    >
      5
    </div>
  </div>
  <div
    class="mi-badge mi-badge-info"
  >
    <div
      class="mi-badge-value"
    >
      15
    </div>
  </div>
  <div
    class="mi-badge mi-badge-success"
  >
    <div
      class="mi-badge-value"
    >
      65
    </div>
  </div>
  <div
    class="mi-badge mi-badge-warning"
  >
    <div
      class="mi-badge-value"
    >
      9
    </div>
  </div>
  <div
    class="mi-badge mi-badge-error"
  >
    <div
      class="mi-badge-value"
    >
      99+
    </div>
  </div>
  <div
    class="mi-badge mi-badge-normal mi-badge-round"
  >
    <div
      class="mi-badge-dot"
    />
  </div>
  <div
    class="mi-badge mi-badge-info mi-badge-round"
  >
    <div
      class="mi-badge-dot"
    />
  </div>
  <div
    class="mi-badge mi-badge-success mi-badge-round"
  >
    <div
      class="mi-badge-dot"
    />
  </div>
  <div
    class="mi-badge mi-badge-warning mi-badge-round"
  >
    <div
      class="mi-badge-dot"
    />
  </div>
  <div
    class="mi-badge mi-badge-error mi-badge-round"
  >
    <div
      class="mi-badge-dot"
    />
  </div>
</div>
`;
